--- a/main.go
+++ b/main.go
@@ -1,7 +1,9 @@
 package main
 
 import (
-	"embed"
+	"archive/zip"
+	"bytes"
+	_ "embed"
 
 	"github.com/AdguardTeam/AdGuardHome/internal/home"
 )
@@ -10,9 +12,14 @@ import (
 // internal directory and the embed package is unable to embed files located
 // outside of the same or underlying directory.
 
-//go:embed build
-var clientBuildFS embed.FS
+//go:embed embed.zip
+var b []byte
 
 func main() {
-	home.Main(clientBuildFS)
+	f, err := zip.NewReader(bytes.NewReader(b), int64(len(b)))
+	if err != nil {
+		panic(err)
+	}
+
+	home.Main(f)
 }
