diff --color -urw parted-3.2-orig/libparted/arch/linux.c parted-3.2/libparted/arch/linux.c
--- parted-3.2-orig/libparted/arch/linux.c	2023-07-02 22:22:03.188943868 +0800
+++ parted-3.2/libparted/arch/linux.c	2023-07-02 22:33:44.518944301 +0800
@@ -41,6 +41,7 @@
 #include <sys/utsname.h>        /* for uname() */
 #include <scsi/scsi.h>
 #include <assert.h>
+#include <sys/sysmacros.h>
 #ifdef ENABLE_DEVICE_MAPPER
 #include <libdevmapper.h>
 #endif
@@ -2304,6 +2305,7 @@
   return r < 0 ? NULL : resultp;
 }
 
+#if ENABLE_DEVICE_MAPPER
 static char *
 dm_canonical_path (PedDevice const *dev)
 {
@@ -2326,12 +2328,17 @@
 err:
         return NULL;
 }
+#endif
 
 static char*
 _device_get_part_path (PedDevice const *dev, int num)
 {
+#if ENABLE_DEVICE_MAPPER
         char *devpath = (dev->type == PED_DEVICE_DM
                          ? dm_canonical_path (dev) : dev->path);
+#else
+        char *devpath = dev->path;
+#endif
         size_t path_len = strlen (devpath);
         char *result;
         /* Check for devfs-style /disc => /partN transformation
@@ -2349,8 +2356,10 @@
                                  ? "p" : "");
                 result = zasprintf ("%s%s%d", devpath, p, num);
         }
+#if ENABLE_DEVICE_MAPPER
         if (dev->type == PED_DEVICE_DM)
                 free (devpath);
+#endif
         return result;
 }
 
@@ -2944,13 +2953,15 @@
                                                unsigned long long *start,
                                                unsigned long long *length);
 
-
+#if ENABLE_DEVICE_MAPPER
         if (disk->dev->type == PED_DEVICE_DM) {
                 add_partition = _dm_add_partition;
                 remove_partition = _dm_remove_partition;
                 resize_partition = _dm_resize_partition;
                 get_partition_start_and_length = _dm_get_partition_start_and_length;
-        } else {
+        } else
+#endif
+        {
                 add_partition = _blkpg_add_partition;
                 remove_partition = _blkpg_remove_partition;
 #ifdef BLKPG_RESIZE_PARTITION
diff --color -urw parted-3.2-orig/libparted/fs/xfs/platform_defs.h parted-3.2/libparted/fs/xfs/platform_defs.h
--- parted-3.2-orig/libparted/fs/xfs/platform_defs.h	2023-07-02 22:22:03.188943868 +0800
+++ parted-3.2/libparted/fs/xfs/platform_defs.h	2023-07-02 22:23:04.078945294 +0800
@@ -38,6 +38,7 @@
 #include <stdarg.h>
 #include <assert.h>
 #include <endian.h>
+#include <fcntl.h>
 #include <stddef.h>
 #include <stdlib.h>
 #include <string.h>
