From fb6056045b4ff88dc6992c4c96457f600d73d35f Mon Sep 17 00:00:00 2001
From: Shuanglei Tao <tsl0922@gmail.com>
Date: Thu, 25 May 2023 14:13:43 +0800
Subject: [PATCH] update default host keys path

---
 default_options.h | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/default_options.h b/default_options.h
index 5132775..535e77b 100644
--- a/default_options.h
+++ b/default_options.h
@@ -21,10 +21,10 @@ IMPORTANT: Some options will require "make clean" after changes */
 /* Default hostkey paths - these can be specified on the command line.
  * Homedir is prepended if path begins with ~/
  */
-#define DSS_PRIV_FILENAME "/etc/dropbear/dropbear_dss_host_key"
-#define RSA_PRIV_FILENAME "/etc/dropbear/dropbear_rsa_host_key"
-#define ECDSA_PRIV_FILENAME "/etc/dropbear/dropbear_ecdsa_host_key"
-#define ED25519_PRIV_FILENAME "/etc/dropbear/dropbear_ed25519_host_key"
+#define DSS_PRIV_FILENAME "/etc/storage/dropbear/dss_host_key"
+#define RSA_PRIV_FILENAME "/etc/storage/dropbear/rsa_host_key"
+#define ECDSA_PRIV_FILENAME "/etc/storage/dropbear/ecdsa_host_key"
+#define ED25519_PRIV_FILENAME "/etc/storage/dropbear/ed25519_host_key"
 
 /* Set NON_INETD_MODE if you require daemon functionality (ie Dropbear listens
  * on chosen ports and keeps accepting connections. This is the default.
-- 
2.34.1

