--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -40,14 +40,6 @@ if(POLICY CMP0075)
        cmake_policy(SET CMP0075 NEW)
 endif()
 
-include(CTest)
-
-if (CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME AND BUILD_TESTING AND
-   (NOT MSVC OR NOT (MSVC_VERSION LESS 1800)) # Tests need at least VS2013
-   )
-add_subdirectory(tests)
-endif()
-
 if (NOT MSVC)  # cmd line apps don't built on Windows currently.
 add_subdirectory(apps)
 endif()
@@ -433,8 +425,6 @@ configure_file(json.h.cmakein ${PROJECT_BINARY_DIR}/json.h @ONLY)
 include_directories(${PROJECT_SOURCE_DIR})
 include_directories(${PROJECT_BINARY_DIR})
 
-add_subdirectory(doc)
-
 # "uninstall" custom target for make generators in unix like operating systems
 # and if that target is not present
 if (CMAKE_GENERATOR STREQUAL "Unix Makefiles")