diff --git a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -41,8 +41,7 @@ include(CheckFunctionExists)
 include(CheckSymbolExists)
 include(CheckCCompilerFlag)

-find_library(PTHREAD_LIBRARY pthread)
-list(APPEND CMAKE_REQUIRED_LIBRARIES ${PTHREAD_LIBRARY})
+list(APPEND CMAKE_REQUIRED_LIBRARIES pthread)

 ############################# MACRO SECTION ############################
 macro(try_c_flag prop flag)